src/gateways/S3-Gateway.ts

```diff
@@ -0,0 +1,70 @@
+import AWS from 'aws-sdk';
+import Metadata from '../interfaces/Metadata';
+
+class S3Gateway {
+  client: AWS.S3;
+  bucketName: string;
+  constructor(client: AWS.S3, bucketName: string) {
+    this.client = client;
+    this.bucketName = bucketName;
+  }
+
+  async create(metadata: Metadata): Promise<Metadata> {
+    await this.client
+      .putObject({
+        Bucket: this.bucketName,
+        Key: `${metadata.documentId}/metadata.json`,
+        Body: Buffer.from(JSON.stringify(metadata)),
+      })
+      .promise();
+
+    return metadata;
+  }
+
+  async createUrl(
+    documentId: string
+  ): Promise<{ url: string; fields: { [key: string]: any } }> {
+    return new Promise((resolve, reject) => {
+      this.client.createPresignedPost(
+        {
+          Bucket: this.bucketName,
+          Expires: 3600,
+          Fields: {
+            success_action_status: '201',
+          },
+          Conditions: [
+            { bucket: this.bucketName },
+            ['starts-with', '$key', `${documentId}/metadata`],
```

Comment:
This sets a condition that all uploaded files must have a key that starts with `${documentId}/metadata`. I don't think that's what we want - I think this can just be `$documentId/` so that all uploaded files must end up in the correct "folder".